### PR TITLE
Can fire guns and secondary attack (right-click) with TK

### PR DESCRIPTION
**Pull request by cacogen**
## About The Pull Request

https://user-images.githubusercontent.com/25089914/158297733-8fb71e73-f79a-4bfe-86a7-9f553db067cb.mp4

https://user-images.githubusercontent.com/25089914/158297718-78f6e819-3c99-452b-9218-4d5cec894d5e.mp4

https://user-images.githubusercontent.com/25089914/158297825-1bc55b1e-f7e1-4b1a-8c9c-569a7a724287.mp4

## Why It's Good For The Game

It's fun

## Changelog

:cl:
expansion: Can fire guns and secondary attack (right-click) with TK. Throwing things with TK now requires throw mode.
/:cl:

### CODE REVIEW

**Comment by BlueMemesauce** [comment]
so warden is fucked?

**Comment by cacogen** [comment]
no

**Comment by KittyNoodle** [comment]
how does this interact with firing pins, because looking at the code it does not look like it interacts at all.
how does this interact with pa rifles?

**Comment by cacogen** [comment]
>how does this interact with firing pins, because looking at the code it does not look like it interacts at all.

no differently

>how does this interact with pa rifles?

nothing happens when you try to fire it with tk

File: code/modules/projectiles/ammunition/_firing.dm [hunk 1]
```diff
@@ -21,11 +21,19 @@
...
 	user.changeNext_move(next_delay)
-
-	user.newtonian_move(get_dir(target, user))
+	if(!tk_firing(user, fired_from))
+		user.newtonian_move(get_dir(target, user))
+	else if(ismovable(fired_from))
+		var/atom/movable/firer = fired_from
+		if(!firer.newtonian_move(get_dir(target, fired_from), instant = TRUE))
+			var/throwtarget = get_step(fired_from, get_dir(target, fired_from))
+			firer.safe_throw_at(throwtarget, 1, 2)
```
Comment (LemonInTheDark) [inline code/modules/projectiles/ammunition/_firing.dm:30]: What does this case cover

Comment (cacogen) [reply code/modules/projectiles/ammunition/_firing.dm:30]: I don't understand the question

Comment (cacogen) [reply code/modules/projectiles/ammunition/_firing.dm:30]: If we can't newtonian move, we're not in space, so we throw ourselves backwards a tile to simulate recoil instead

Comment (LemonInTheDark) [reply code/modules/projectiles/ammunition/_firing.dm:30]: what is fired_from in this case?

Comment (cacogen) [reply code/modules/projectiles/ammunition/_firing.dm:30]: fired_from is the atom that fired the bullet

Comment (LemonInTheDark) [reply code/modules/projectiles/ammunition/_firing.dm:30]: aight, thanks

**Comment by CMDR-Gungnir** [comment]
This sounds like a horrible idea and I can already see the ungodly amount of grief this is going to cause (let alone any possible interactions between Hulk and TK). TK already has enough Shitters and Griefers using it, I don't think we need to arm them, too.

**Comment by cacogen** [comment]
Why is it that I made this exciting new feature previously only dreamed of and the only people who seem to be interested are joyless and complaining about balance

**Comment by CMDR-Gungnir** [comment]
> Why is it that I made this exciting new feature previously only dreamed of and the only people who seem to be interested are joyless and complaining about balance

Because it's a very cool idea on paper, but the idea of giving griefing shitters more tools to grief with sounds like a nightmare. Don't get me wrong, I think it's extremely cool and I like it as an idea.

I just know how it's going to get used will make it extremely un-cool.

**Comment by Valorium-RB** [comment]
TK is barely usable as is. As a frequent Warden player, +1 from me.

Also, blindfolds, flashbangs and flashes exist for a reason. They can't TK what they can't see.

**Comment by Serxule** [comment]
> Person with TK gets arrested, sees a loose disabler/laser gun on the desk, proceeds to stamcrit warden and two officers. This is how you get people with TK treated like Hulks: uncontainable. Cool concept, not sure if it's a good call to be buffing genetics in it's current state.

they said firing pins wouldnt change at all, so i think hulks would still be unable to use them

**Comment by PKPenguin321** [comment]
>Why is it that I made this exciting new feature previously only dreamed of and the only people who seem to be interested are joyless and complaining about balance

It always happens. People didn't want to add carrot shivs of all things because they "would be OP"
TK is fun but really really useless as is so this is a good change

**Comment by CMDR-Gungnir** [comment]
> > Why is it that I made this exciting new feature previously only dreamed of and the only people who seem to be interested are joyless and complaining about balance
> 
> It always happens. People didn't want to add carrot shivs of all things because they "would be OP" TK is fun but really really useless as is so this is a good change

TK is useless for anything except griefing, yeah. But that's the part that makes me afraid. It's just giving the griefing shitters more tools.

**Comment by PKPenguin321** [comment]
>TK is useless for anything except griefing, yeah.

Shut up, dude.

**Comment by CMDR-Gungnir** [comment]
> > TK is useless for anything except griefing, yeah.
> 
> Shut up, dude.

This just tells me I'm right.

**Comment by ATH1909** [comment]
on the one hand this'll make it arbitrarily harder to move guns around with tk if you just want to, say, pull the gun towards yourself and pick it up or yank it away from an enemy combatant, and someone will eventually end up accidentally shooting themselves because they didn't know about this functionality

on the other hand fuck that's cool

don't we have a recoil variable or something for guns? maybe we should make the number of tiles the gun is knocked back depend on that, as I don't think that energy guns and the like actually have any recoil to be moved by

would it be possible to add a way to move the gun the way you'd normally move a TK'd item?

in that last mp4, it looks like the welder ended up hitting the window after repairing it. you might want to look into that.

**Comment by cacogen** [comment]
>would it be possible to add a way to move the gun the way you'd normally move a TK'd item?

You use throw mode to throw things with TK now, which I figured out how to implement in what I personally think is an ingenious way.

>don't we have a recoil variable or something for guns? maybe we should make the number of tiles the gun is knocked back depend on that, as I don't think that energy guns and the like actually have any recoil to be moved by

That's a good idea, I'll look into that.

>in that last mp4, it looks like the welder ended up hitting the window after repairing it. you might want to look into that.

I think that should be fixed now, but I'll make sure of it.

>what happens if you try to hold someone at gunpoint with tk?

I haven't tested it outside of attempting to hold myself up so I don't know if it'll work without proximity or what will happen if the user moves but the gun maintains proximity, for example.

**Comment by alphanerdd** [comment]
this feels like the kind of thing telekinesis should be able to do already, people complaining are just scared of the genetide

**Comment by Fikou** [comment]
the throwing change doesnt seem to be included anywhere but the changelog. its a pretty big change to tk overall so you should probably add instructions to it ingame

**Comment by cacogen** [comment]
if even players can't press the up thumb on something this good it's never getting merged anyway

**Comment by Shadowflame909** [comment]
Genetics powers were all useless and underused lately. This is a much needed shift in the other direction. Awesome Stuff Cacogen!

**Comment by LemonInTheDark** [comment]
I don't have a lot of confidence in our players to not be dickweeds with this honestly

**Comment by cacogen** [comment]
It should always have worked like this and probably didn't due to the limitations of the time. If it causes problems, it can be nerfed by increasing the cooldown on firing or making it less accurate.

**Comment by ShizCalev** [comment]
i love this idea.

i actually wrote very similar code years ago.

then, as someone who used to genetide, i asked myself if we _should_ give players who like to genetide **this much** power if they have xray+tk.

no from me.

**Comment by Shadowflame909** [comment]
> i love this idea.
> 
> i actually wrote very similar code years ago.
> 
> then, as someone who used to genetide, i asked myself if we _should_ give players who like to genetide **this much** power if they have xray+tk.
> 
> no from me.

Xray got  made admin only years ago shiv

**Comment by Eronymun** [comment]
If the gun firing change doesn't pass for balance reasons, I'd still love to see the right click and throw rework as they have more practical applications in engineering.

**Comment by LemonInTheDark** [comment]
lmao shiz

**Pull request merged**